### ISSUE TITLE
ease the effort to override factories

### ISSUE THREAD

**Issue by schleichardt** [commercetools/commercetools-sunrise-java#416]
- [x] AddAddressPageContentFactory
- [x] AddressBeanFactory
- [x] AddressBookPageContentFactory
- [x] AddressFormBeanFactory
- [x] AddressFormSettingsFactory
- [x] AuthenticationPageContentFactory
- [x] BreadcrumbBeanFactory
- [x] CartDetailPageContentFactory
- [x] CartLikeBeanFactory
- [x] ChangeAddressPageContentFactory
- [x] CheckoutAddressFormSettingsFactory
- [x] CheckoutAddressPageContentFactory
- [x] CheckoutConfirmationPageContentFactory
- [x] CheckoutShippingPageContentFactory
- [x] CheckoutThankYouPageContentFactory
- [x] CountryFormFieldBeanFactory
- [x] FacetedSearchSelectorFactory
- [x] FacetedSearchSelectorListFactory
- [x] LineItemBeanFactory
- [x] LineItemsBeanFactory
- [x] MyOrderDetailPageContentFactory
- [x] MyOrderListPageContentFactory
- [x] MyPersonalDetailsPageContentFactory
- [x] OrderOverviewBeanFactory
- [x] PageMetaFactory
- [x] PageMetaFactoryImpl
- [x] PageNavMenuFactory
- [x] PageNavMenuFactoryImpl
- [x] PaginationBeanFactory
- [x] PaginationFactory
- [x] ProductAttributeBeanFactory
- [x] ProductAttributeBeanFactoryInjectless
- [x] ProductBeanFactory
- [x] ProductDetailPageContentFactory
- [x] ProductListBeanFactory
- [x] ProductOverviewPageContentFactory
- [x] ProductsPerPageSelectorFactory
- [x] ProductVariantBeanFactory
- [x] ProductVariantBeanFactoryInjectless
- [x] SearchBoxFactory
- [x] SelectFacetedSearchSelectorFactory
- [x] SortSelectorFactory
- [x] SunriseDataBeanFactory
- [x] TitleFormFieldBeanFactory
## \* [ ] remove WithOverridablePageData

a suggestion:

Factories implement this interface:

``` java
public interface OverridableBeanFactory<T> {
    T createNewBean();
}
```

For example:

``` java
public class ProductListBeanFactory implements OverridableBeanFactory<ProductListBean> {

    @Inject
    @Named("new")
    private CategoryTree categoryTreeInNew;
    @Inject
    private ProductBeanFactory productBeanFactory;

    public ProductListBean create(final Iterable<ProductProjection> products) {
        final ProductListBean bean = createNewBean();
//details
        return bean;
    }

    private ProductThumbnailBean createThumbnail(final ProductProjection product, final ProductVariant variant) {
        final ProductThumbnailBean bean = new ProductThumbnailBean();
//details
        return bean;
    }

    private static void fillIsSaleInfo(final ProductThumbnailBean bean, final ProductBean productBean) {
        final boolean isSale = productBean != null && productBean.getVariant() != null && productBean.getVariant().getPriceOld() != null;
        bean.setSale(isSale);
    }

    @Override
    public ProductListBean createNewBean() {
        return new ProductListBean();
    }
}
```

Given a bean subclass

``` java
public class MyProductListBean extends ProductListBean {
    private Integer productCount;

    public Integer getProductCount() {
        return productCount;
    }

    public void setProductCount(final Integer productCount) {
        this.productCount = productCount;
    }
}
```

then the subclass of the factory would look like this:

``` java
public class MyProductListBeanFactory extends ProductListBeanFactory {
    @Override
    public ProductListBean createNewBean() {//this method us used by the base class to create the bean instance
        return new MyProductListBean();
    }

    @Override
    public ProductListBean create(final Iterable<ProductProjection> products) {
        final List<ProductProjection> productList = new LinkedList<>();
        products.forEach(productList::add);
        final MyProductListBean productListBean = (MyProductListBean) super.create(productList);//the solution is a cast
        productListBean.setProductCount(productList.size());//set the new field
        return productListBean;
    }
}
```

I think other solutions would have more layers of abstraction (type parameters and using interfaces).

**Comment by schleichardt** [comment]
alternative proposal:

```
public class MyProductBean extends ProductBean {
    private Integer count;

    public Integer getCount() {
        return count;
    }

    public void setCount(final Integer count) {
        this.count = count;
    }
}
```

```
public class MyProductBeanFactory extends ProductBeanFactory {
    @Override
    public ProductBean create(final ProductProjection product, final ProductVariant variant) {
        final MyProductBean myProductBean = fill(product, variant, new MyProductBean());
        myProductBean.setCount(product.getAllVariants().size());
        return myProductBean;
    }
}
```

```
public class ProductBeanFactory extends Base {
    public ProductBean create(final ProductProjection product, final ProductVariant variant) {
        return fill(product, variant, new ProductBean());
    }

    protected <T extends ProductBean> T fill(final ProductProjection product, final ProductVariant variant, final T productBean) {
        fillIds(product, variant, productBean);
        fillDescription(product, productBean);
        fillGallery(variant, productBean);
        fillDetails(variant, productBean);
        fillVariant(product, variant, productBean);
        fillVariantIdentifiers(productBean);
        fillVariants(product, productBean);
        fillAttributes(product, variant, productBean);
        return productBean;
    }
}
```

**Comment by schleichardt** [comment]
@lauraluiz do we need the customer parameter in com.commercetools.sunrise.myaccount.addressbook.addaddress.AddAddressPageContentFactory, it is unused

**Comment by schleichardt** [comment]
there is also an unused parameter customer in com.commercetools.sunrise.myaccount.addressbook.changeaddress.ChangeAddressPageContentFactory#create

**Comment by lauraluiz** [comment]
Solved since v1.0.0-M1.
